text/0076-entity-slice-validation.md

```diff
@@ -0,0 +1,443 @@
+
+All of the Cedar example policy sets validate with level-based validation, either at level 1 or 2; see the Appendix for details.
+
+## Motivation
+
+Currently, Cedar provides no support for slicing. Users are required to either be inefficient and produce their entire entity store as a slice, or derive their own ad-hoc slicing algorithm. We'd like to provide an out-of-the-box solution that addresses many user's needs. Entity slicing in general is a problem very tied to the specifics of application/deployment. It's likely impossible to have a fully general solution. This RFC is attempting to provide a solution for entity slicing that works for the 80% case, while acknowledging that it won't work for everyone. 
+
+Additionally, there are some application scenarios where the actors writing the policies and the people performing slicing are not the same people, and thus the slicers need to impose some kind of restrictions on policy authors.
+
+This RFC has a close connection to [RFC 74: Entity Slicing using Data Tries](https://github.com/cedar-policy/rfcs/pull/74), which also seeks to provide a sound foundation for entity slicing. We discuss the relationship between them in the Alternatives below; ultimately, we believe they are complementary and should coexist.
+
+## Detailed design
+
+### Entity dereferences
+
+An "Entity Dereference" is any operation that requires directly accessing an entity’s data. The following is an exhaustive list of entity dereferencing operations
+
+1. `e1 in e2` dereferences `e1` (but not `e2`)
+2. `e1.foo` dereferences `e1` iff `e1` is an entity
+3. `e1 has foo` dereferences `e1` iff `e1` is an entity
+
+Note that testing equality *is not* a dereferencing operation. Therefore expressions like `e1 == e2`, `e1.contains(e2)` and `e3 in e2` do not dereference `e1` or `e2`.
+
+Also note that by “dereference” above, we are only referring to the given operation, not about any operations that might be executed to evaluate subexpressions. For example, while `e1 in e2` only dereferences `e1` (per rule 1 above), if `e2` is `principal.foo`, then of course evaluating `e2` dereferences `principal` (per rule 2 above).
+
+### Entity roots
+
+The variables `principal`, `action`, and `resource` are *entity roots*. (Entity literals are *not* entity roots, and operations on such literals are limited, as described below.) They directly name entities from a request that could be dereferenced in a policy. A `context` is not an entity root directly; rather, any attributes it (transitively) contains that have entity type are considered roots. For example, consider the following schema:
```

Comment:
```suggestion
The variables `principal`, `action`, and `resource` are *entity roots*. (Entity literals are *not* entity roots, and operations on such literals are limited, as described below. Slots and the `Unknown` entity are treated as entity literals because slots will be replaced by literals and `Unknown` cannot be dereferenced.) They directly name entities from a request that could be dereferenced in a policy. A `context` is not an entity root directly; rather, any attributes it (transitively) contains that have entity type are considered roots. For example, consider the following schema:
```